src/app/reporting/LoggingManagement.cpp

```diff
@@ -0,0 +1,1306 @@
+ * corresponds to the least important events.
+ *
+ * @param[in] inMgr         ExchangeManager to be used with this logging subsystem
+ *
+ * @param[in] inNumBuffers  Number of elements in inLogStorageResources array
+ *
+ * @param[in] inLogStorageResources  An array of LogStorageResources for each priority level.
+ *
+ * @note This function must be called prior to the logging being used.
+ */
+void LoggingManagement::CreateLoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                                const LogStorageResources * const apLogStorageResources)
+{
+    new (&sInstance) LoggingManagement(apMgr, aNumBuffers, apLogStorageResources);
+}
+
+/**
+ * @brief Perform any actions we need to on shutdown.
+ */
+void LoggingManagement::DestroyLoggingManagement(void)
+{
+    Platform::CriticalSectionEnter();
+    sInstance.mState        = kLoggingManagementState_Shutdown;
+    sInstance.mpEventBuffer = NULL;
+    Platform::CriticalSectionExit();
+}
+
+/**
+ * @brief Set the ExchangeManager to be used with this logging subsystem.  On some
+ *   platforms, this may need to happen separately from CreateLoggingManagement() above.
+ *
+ * @param[in] inMgr         ExchangeManager to be used with this logging subsystem
+ */
+CHIP_ERROR LoggingManagement::SetExchangeManager(Messaging::ExchangeManager * apMgr)
+{
+    mpExchangeMgr = apMgr;
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * @brief
+ *   LoggingManagement constructor
+ *
+ * Initialize the LoggingManagement with an array of LogStorageResources.  The
+ * array must provide a resource for each valid priority level, the elements
+ * of the array must be in increasing numerical value of priority (and in
+ * increasing priority); the first element in the array corresponds to the
+ * resources allocated for least important events, and the last element
+ * corresponds to the most critical events.
+ *
+ * @param[in] apMgr         ExchangeManager to be used with this logging subsystem
+ *
+ * @param[in] aNumBuffers  Number of elements in inLogStorageResources array
+ *
+ * @param[in] apLogStorageResources  An array of LogStorageResources for each priority level.
+ *
+ */
+
+LoggingManagement::LoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                     const LogStorageResources * const apLogStorageResources)
+{
+    CircularEventBuffer * current = NULL;
+    CircularEventBuffer * prev    = NULL;
+    CircularEventBuffer * next    = NULL;
+    size_t i, j;
+
+    VerifyOrDie(aNumBuffers > 0);
+
+    mpExchangeMgr = apMgr;
+
+    for (j = 0; j < aNumBuffers; j++)
+    {
+        i = aNumBuffers - 1 - j;
+
+        next = (i > 0) ? static_cast<CircularEventBuffer *>(apLogStorageResources[i - 1].mpBuffer) : NULL;
+
+        VerifyOrDie(apLogStorageResources[i].mBufferSize > sizeof(CircularEventBuffer));
+
+        new (apLogStorageResources[i].mpBuffer)
```

Comment:
Is placement new guaranteed to be correctly aligned?

I find this style hard to review and maintain. Could we pass in pointers instead (e.g. pass in CircularEventBuffers arrays and not just flat data buffers)?

Reply:
Looking at the circularEventBuffer class, it seems that it contains typese that reuqire alignment (pointers), so this is specifically not type safe. This explains the previous usage of uint64_t instead of pure bytes ... but that seems very fragile.

Please refactor this code to not use placement new anymore.

Reply:
Removed placement new. Done, thanks